### PR TITLE
Adding support for FIDO/FIDO2 APIs

### LINKED ISSUE #1148: Unsupported browser message when security key is connected
### Description: 
Unsupported browser message when security key is connected

### Steps to Reproduce 
  1. Visit a page that supports security key authentication
  2. Connect the security key when it asks to authenticate
  3. Disconnect device from network
  4. Validate using the security key, page load fails
  5. With security key connected reload the page
  6. Page shows security key unsupported 

**Actual result:** 
Unsupported browser message when security key is connected

**Expected result:**
Should just ask for security key authentication again 

**Reproduces how often:** [Easily reproduced, Intermittent Issue]
Easy

**Device details:** 
iPhone 7+ (iOS 12.3.1)

**Website problems only:**
- did you check with Brave Shields down?
- did you check in Safari/Firefox (WkWebView-based browsers)?

### LINKED ISSUE #1150: Authentication doesn't work when security key is connected and disconnected quickly
### Description: 
Authentication doesn't work when security key is connected and disconnected quickly

### Steps to Reproduce 
  1. Login to a site that supports security key
  2. Wait for the message to use security key
  3. Quickly connect the security key and disconnect
  4. Site doesn't authenticate your account
  5. Reconnect the security key and try to authenticate, account never authenticates
https://bravesoftware.slack.com/archives/CKH2AUS6S/p1560432037010500

**Actual result:** 
Authentication doesn't work when security key is connected and disconnected quickly

**Expected result:**
Should be able to authenticate to the site

**Reproduces how often:** [Easily reproduced, Intermittent Issue]
Easy

**Device details:** 
iPad Pro running 12.2 and iPhone running 12.3.1 

**Website problems only:**
- did you check with Brave Shields down?
- did you check in Safari/Firefox (WkWebView-based browsers)?

### LINKED ISSUE #1159: Add support for U2F/WebAuthn hardware security keys on iOS
### Test Plan

Specified here: https://github.com/brave/brave-ios/pull/1158

### LINKED ISSUE #1163: Security keys don't work with login.gov
### Description: 
Adding security keys on login.gov fails with errors.

### Steps to Reproduce 
  1. Navigate to login.gov and try to add a security key
  2. Fails!

**Actual result:** 
Fails

**Expected result:**
Should be successful

**Reproduces how often:** [Easily reproduced, Intermittent Issue]
Easily

**Brave Version:** 
1.10

### LINKED ISSUE #1176: FIDO/FIDO2 APIs fail on iPad Pro
### Description: 
FIDO/FIDO2 APIs should work with iPad Pro.

### Steps to Reproduce 
  1. Navigate to demo.yubico.com/u2f
  2. Initiate the U2F workflow
  3. Fails

**Actual result:** 
Fails

**Expected result:**
Should succeed.

**Reproduces how often:** [Easily reproduced, Intermittent Issue]
Easily

**Device details:** 
iPad Pro

### PR DESCRIPTION
**Pull request by jumde**
fix https://github.com/brave/brave-ios/issues/1148
fix https://github.com/brave/brave-ios/issues/1150
fix https://github.com/brave/brave-ios/issues/1159
fix https://github.com/brave/brave-ios/issues/1163
fix https://github.com/brave/brave-ios/issues/1176

## Description
This PR implements javascript FIDO/FIDO2 APIs for the Brave iOS browser. With these APIs users will be able to use the lightning yubikeys keys as a second factor for authentication.

FIDO2 (WebAuthn) and FIDO rely on javascript APIs (navigator.credentials.create, navigator.credentials.get, u2f.sign, u2f.register). These are currently not available with WKWebView.

Scripts which define these APIs are injected at document start and end. These APIs hooks invoke services provided by the Yubico SDK to sign and register the requests passed using javascript.

For more details about the APIs refer to:
https://fidoalliance.org/fido2/
https://FIDOalliance.org/specs/fido-u2f-v1.0-nfc-bt-amendment-20150514/fido-u2f-javascript-api.html

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.

## Test Plan:

1. Navigate to the demo sites and make sure FIDO/FIDO2 workflow completes.
2. For other real sites:
  - create an account
  - enable 2FA
  - Registering 2FA keys should succeed.
  - Logout and try logging back in with the security key
  - Should succeed.

### Demo Sites:

1. WebAuthn
  - demo.yubico.com/u2f
  - webauthn.io
  - webauthntest.azurewebsites.net/
2. High Level Legacy U2F - mdp.github.io/u2fdemo
3. Low Level Legacy U2F - u2fdemo.hypersecu.com
4. APIs in secure context - http://dustiest-limitation.000webhostapp.com/nonsecure-u2f.html
5. APIs in 3p iframes - http://dustiest-limitation.000webhostapp.com/fido-iframe.html

### Sites that are working:
Without UA change:

1. github.com
2. bitbucket.org
3. 1Password.com
4. login.gov

Need UA change: https://github.com/brave/brave-ios/pull/1158/files#diff-1b680c574387c7d713ce005b70d406acR58

1. console.aws.amazon.com
2. twitter.com

## Reviewer Checklist:

- [x] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [x] Issues are assigned to at least one epic.
- [x] Issues include necessary QA labels:
  - [x] `QA/(Yes|No)`
  - [x] `release-notes/(include|exclude)`
  - [x] `bug` / `enhancement`
- [x] Necessary security reviews have taken place.
- [x] Adequate test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable)

### CODE REVIEW

File: Shared/UserAgent.swift [hunk 1]
```diff
@@ -55,7 +55,8 @@ open class UserAgent {
      */
     public static func defaultUserAgent() -> String {
         assert(Thread.current.isMainThread, "This method must be called on the main thread.")
-
+        return "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.157 Safari/537.36"
```
Comment (jumde) [inline Shared/UserAgent.swift]: - [x] Revert before merge

File: bootstrap.sh [hunk 2]
```diff
@@ -64,3 +64,23 @@ do
...
+
+# Build Yubikit
+YUBIKIT_DIR=Carthage/Checkouts/yubikit-ios
+YUBIKIT_OUT=ThirdParty/YubiKit
+
+rm -rf YUBIKIT_DIR
+rm -rf YUBIKIT_OUT
+
+if [ ! -d "$YUBIKIT_DIR" ]; then
+  git clone https://github.com/jumde/yubikit-ios/ $YUBIKIT_DIR
```
Comment (kylehickinson) [inline bootstrap.sh]: Should be doing a shallow clone here, since we don't need any git history for a third party lib (`git clone --depth 1 <url> <dir>`)

File: bootstrap.sh [hunk 3]
```diff
@@ -64,3 +64,23 @@ do
...
+fi
+
+SRCDIR=$PWD
+mkdir -p $SRCDIR/$YUBIKIT_OUT/include
+cd $YUBIKIT_DIR/
+git checkout simulator_release
+cd YubiKit
+sh build.sh
```
Comment (kylehickinson) [inline bootstrap.sh]: This prints out too much output and Travis hits the maximum log level before the build completes. Either:
- Write to a log file and if an error occurs, print log file destination. (`sh build.sh > file.log 2>&1`)
- Dump all stdout to /dev/null (`sh build.sh > /dev/null`), allowing only errors to surface in `stderr`

File: bootstrap.sh [hunk 4]
```diff
@@ -64,3 +64,23 @@ do
...
+rm -rf YUBIKIT_DIR
+rm -rf YUBIKIT_OUT
+
+if [ ! -d "$YUBIKIT_DIR" ]; then
+  git clone https://github.com/jumde/yubikit-ios/ $YUBIKIT_DIR
+fi
+
+SRCDIR=$PWD
+mkdir -p $SRCDIR/$YUBIKIT_OUT/include
+cd $YUBIKIT_DIR/
```
Comment (kylehickinson) [inline bootstrap.sh]: You can utilize `pushd` and `popd` here instead of `cd`ing into YUBIKIT_DIR and cd'ing back to src:

```
pushd $YUBIKIT_DIR
# pwd is now $YUBIKIT_DIR
popd
# pwd is now whatever it was before pushd was called
```

File: bootstrap.sh [hunk 5]
```diff
@@ -64,3 +64,23 @@ do
     && cp -n Local.templates/$CONFIG_FILE Local/$CONFIG_FILE \
   )
 done
+
+# Build Yubikit
+YUBIKIT_DIR=Carthage/Checkouts/yubikit-ios
+YUBIKIT_OUT=ThirdParty/YubiKit
+
+rm -rf YUBIKIT_DIR
```
Comment (kylehickinson) [inline bootstrap.sh]: 1. These lines are not evaluating to the value because it is not prefixed with `$`
2. These two lines will error if the directories don't exist. If we add `set -x` to the top of this file (which will happen eventually), it would stop here.

Directory should be checked that it exists before it is attempted to be removed:

```
[ -d "$YUBIKIT_DIR" ] && rm -rf "$YUBIKIT_DIR"
```

**Review by kylehickinson — CHANGES_REQUESTED** [review]

File: bootstrap.sh (hunk 2 above)
Comment (kylehickinson) [reply bootstrap.sh]: Also, if you are going to checkout immediately after to a different branch, you should clone it just to that branch:

`git clone -b simulator_release --single-branch --depth 1 https://github.com/jumde/yubikit-ios/ $YUBIKIT_DIR`

Comment (jumde) [reply bootstrap.sh]: resolved!

File: bootstrap.sh (hunk 3 above)
Comment (jumde) [reply bootstrap.sh]: resolved!

File: bootstrap.sh [hunk 6]
```diff
@@ -64,3 +64,21 @@ do
...
+SRCDIR=$PWD
+
+rm -rf $YUBIKIT_DIR
+rm -rf $YUBIKIT_OUT
+
+git clone -b simulator_release --single-branch --depth 1 https://github.com/jumde/yubikit-ios/ $YUBIKIT_DIR
+
+mkdir -p $SRCDIR/$YUBIKIT_OUT/include
+pushd $YUBIKIT_DIR/
+git checkout simulator_release
```
Comment (kylehickinson) [inline bootstrap.sh]: This is no longer required, since you cloned only this branch :)

**Review by kylehickinson — DISMISSED** [review]
Sorry one last small thing. Also I dont see the change for making sure the directories are there before removing them?

```shell
[ -d "$YUBIKIT_DIR" ] && rm -rf "$YUBIKIT_DIR"
[ -d "$YUBIKIT_OUT" ] && rm -rf "$YUBIKIT_OUT"
```

File: bootstrap.sh (hunk 6 above)
Comment (jumde) [reply bootstrap.sh]: yes, I missed this earlier. Should be fixed now.

File: bootstrap.sh (hunk 5 above)
Comment (jumde) [reply bootstrap.sh]: > Sorry one last small thing. Also I dont see the change for making sure the directories are there before removing them?
> 
> ```shell
> [ -d "$YUBIKIT_DIR" ] && rm -rf "$YUBIKIT_DIR"
> [ -d "$YUBIKIT_OUT" ] && rm -rf "$YUBIKIT_OUT"
> ```

I don't think we need to check if the directory exists, `rm` will do nothing if the directory does not exist. Let me know if you have any concerns.

Comment (kylehickinson) [reply bootstrap.sh]: Sorry, didn't realize that `-f` overrides this and does not fail if the directory doesn't exist. Should be good then

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift [hunk 7]
```diff
@@ -247,6 +247,12 @@ extension BrowserViewController: WKNavigationDelegate {
                 }
             }
         }
+        
+        // U2F code below
+        if let tab = tabManager[webView] {
+            // U2F should only be available in secure contexts
+            tab.userScriptManager?.isU2FEnabled = webView.hasOnlySecureContent
```
Comment (kylehickinson) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift]: FYI, not sure if its better here, but we have KVO setup here:

https://github.com/brave/brave-ios/blob/38554971754cace94437d7aa8e98defcd61653ce/Client/Frontend/Browser/BrowserViewController.swift#L952-L961

Which you could use to update the `userScriptManager` as the `hasOnlySecureContent` changes on the fly

Would need to test to make sure its available in time

File: Client/WebAuthN/WebAuthnClientData.swift [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    case create = "webauthn.create"
+    case get = "webauthn.get"
+}
+
+struct WebAuthnClientData {
+    var type: String
+    var challenge: String
+    var origin: String
+    
+    enum CodingKeys: String, CodingKey {
```
Comment (kylehickinson) [inline Client/WebAuthN/WebAuthnClientData.swift:17]: Redundant: CodingKeys here is synthesized by the compiler with the property names and since you didn't change their keys it will be the same

File: Client/WebAuthN/WebAuthnClientData.swift [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        case type
+        case challenge
+        case origin
+    }
+    
+    init(type: String, challenge: String, origin: String) {
```
Comment (kylehickinson) [inline Client/WebAuthN/WebAuthnClientData.swift]: Redundant: This is synthesized for all Swift structs

File: Client/WebAuthN/WebAuthnClientData.swift [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        try container.encode(type, forKey: .type)
+        try container.encode(origin, forKey: .origin)
+        
+        let challengeData = Data(base64Encoded: challenge)
+        let websafeChallenge = challengeData?.websafeBase64String()
+        try container.encode(websafeChallenge, forKey: .challenge)
+    }
+}
+
+func clientDataHash (data: Data) -> Data? {
```
Comment (kylehickinson) [inline Client/WebAuthN/WebAuthnClientData.swift:36]: I don't see this used anywhere?

File: Client/WebAuthN/WebAuthnRegisterRequest.swift [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    var rpID: String?
+    var rpName: String
+    
+    var pubKeyAlg: Int
+    var residentKey: Bool
+    
+    var userVerification: Bool
+    var challenge: String
+    
+    enum CodingKeys: String, CodingKey {
```
Comment (kylehickinson) [inline Client/WebAuthN/WebAuthnRegisterRequest.swift]: This doesn't seem to be used

File: Client/WebAuthN/WebAuthnRegisterRequest.swift [hunk 12]
```diff
@@ -0,0 +1,90 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+import Shared
+
+struct WebAuthnRegisterRequest {
```
Comment (kylehickinson) [inline Client/WebAuthN/WebAuthnRegisterRequest.swift]: Overall I'd say this is a bit hard to follow, and I'd recommend smaller structs, with compiler synthesized `Codable`, based on the JSON structure rather than a single flat struct, unless you have a specific reason for it?

You only seem to do anything custom with the `pubKeyAlg` and possibly the `AuthenticatorSelection` so it should be simple.

Example:

```swift
struct WebAuthnRegisterRequest: Decodable {
  struct PublicKey: Decodable {
    struct AuthenticatorSection: Decodable {
      let requireResidentKey: Bool
      let userVerification: String
    }
    struct PublicKeyCredentialParameters: Decodable {
      var alg: String
      var type: String
    }
    struct User: Decodable {
      var displayName: String
      var name: String
      var id: String
    }
    let attestation: String
    let authenticatorSection: AuthenticatorSection?
    let pubKeyCredParams: [PublicKeyCredentialParameters]
    ...
    ...
    let user: User
  }
  let publicKey: PublicKey
}
```

File: ClientTests/U2FTests.swift [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+            let request =  try JSONDecoder().decode(WebAuthnRegisterRequest.self, from: jsonData)
+            XCTAssertEqual(request.username, "Brave demo user", "request username is correct.")
+            XCTAssertEqual(request.userID, "OvQO5490o1w89Op/9dp4w7VvKuLEk5NHcfOnc2ZECtc=", "request user id is correct.")
+            XCTAssertEqual(request.rpID, expectedrpID, "request rp id is correct.")
+            XCTAssertEqual(request.rpName, "Brave", "request rp name is correct.")
+            XCTAssertEqual(request.pubKeyAlg, -7, "request pub key alg is correct.")
+            XCTAssertFalse(request.residentKey, "request resident key is correct.")
+            XCTAssertEqual(request.challenge, "mdMbxTPACurawWFHqkoltSUwDear2OZQVl/uhBNqiaM=", "request challenge is correct.")
+        } catch {
+            XCTFail()
```
Comment (kylehickinson) [inline ClientTests/U2FTests.swift]: May as well pass the implicit error in the failure: `XCTFail("\(error)")`

File: ClientTests/U2FTests.swift [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+        }
+        
+        do {
+            let request =  try JSONDecoder().decode(WebAuthnAuthenticateRequest.self, from: jsonData)
+            XCTAssertEqual(request.rpID, expectedrpID, "request rp id is correct.")
+            XCTAssertEqual(request.challenge, "mdMbxTPACurawWFHqkoltSUwDear2OZQVl/uhBNqiaM=", "request challenge is correct.")
+            XCTAssertEqual(request.allowCredentials.count, 1, "request allowCredential count is correct")
+            XCTAssertEqual(request.allowCredentials.first, "OvQO5490o1w89Op/9dp4w7VvKuLEk5NHcfOnc2ZECtc=", "request allowCredential is correct")
+        } catch {
+            XCTFail()
```
Comment (kylehickinson) [inline ClientTests/U2FTests.swift]: Same as above

File: ClientTests/U2FTests.swift [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+@testable import Client
+
+class U2FTests: XCTestCase {
+    
+    let expectedrpID = "demo.brave.com"
+    
+    func testWebAuthnRegisterRequest() {
+        let webAuthnRegisterData = "{\"publicKey\":{\"attestation\":\"direct\",\"authenticatorSelection\":{\"requireResidentKey\":false,\"userVerification\":\"discouraged\"},\"pubKeyCredParams\":[{\"alg\":-7,\"type\":\"public-key\"},{\"alg\":-257,\"type\":\"public-key\"}],\"rp\":{\"id\":\"demo.brave.com\",\"name\":\"Brave\"},\"timeout\":90000,\"challenge\": \"mdMbxTPACurawWFHqkoltSUwDear2OZQVl/uhBNqiaM=\",\"user\":{\"displayName\":\"Brave demo user\",\"name\":\"Brave demo user\",\"id\":\"OvQO5490o1w89Op/9dp4w7VvKuLEk5NHcfOnc2ZECtc=\"},\"excludeCredentials\":[]},\"signal\":{}}"
+        guard let jsonData = webAuthnRegisterData.data(using: String.Encoding.utf8) else {
+            XCTFail()
```
Comment (kylehickinson) [inline ClientTests/U2FTests.swift]: Should provide a message with the failure for CI

File: ClientTests/U2FTests.swift [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+            XCTAssertEqual(request.challenge, "mdMbxTPACurawWFHqkoltSUwDear2OZQVl/uhBNqiaM=", "request challenge is correct.")
+        } catch {
+            XCTFail()
+        }
+    }
+    
+    func testWebAuthnAuthenticateRequest() {
+        let webAuthnAuthenticateData = "{\"publicKey\":{\"allowCredentials\":[{\"type\":\"public-key\",\"id\":\"OvQO5490o1w89Op/9dp4w7VvKuLEk5NHcfOnc2ZECtc=\"}],\"rpId\":\"demo.brave.com\",\"timeout\":90000,\"userVerification\":\"discouraged\",\"challenge\":\"mdMbxTPACurawWFHqkoltSUwDear2OZQVl/uhBNqiaM=\"},\"signal\":{}}"
+        guard let jsonData = webAuthnAuthenticateData.data(using: String.Encoding.utf8) else {
+            XCTFail()
```
Comment (kylehickinson) [inline ClientTests/U2FTests.swift]: Should provide a message with the failure for CI

File: ClientTests/U2FTests.swift [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+            XCTAssertEqual(request.allowCredentials.count, 1, "request allowCredential count is correct")
+            XCTAssertEqual(request.allowCredentials.first, "OvQO5490o1w89Op/9dp4w7VvKuLEk5NHcfOnc2ZECtc=", "request allowCredential is correct")
+        } catch {
+            XCTFail()
+        }
+    }
+    
+    func testValidateRPId() {
+        guard let currentURL = URL(string: "https://rp.example.domain") else {
+            XCTFail()
```
Comment (kylehickinson) [inline ClientTests/U2FTests.swift]: Should provide a message with the failure for CI

File: Client/WebAuthN/WebAuthnAuthenticateRequest.swift [hunk 18]
```diff
@@ -0,0 +1,47 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+import Shared
+
+struct WebAuthnAuthenticateRequest {
+    var rpID: String?
+    var challenge: String
+    var allowCredentials: Array<String> = []
```
Comment (kylehickinson) [inline Client/WebAuthN/WebAuthnAuthenticateRequest.swift]: Usually written like `[String]` instead of `Array<String>`

File: Client/U2FExtensions.swift [hunk 19]
```diff
@@ -0,0 +1,893 @@
...
+                // The reponse from the key must not be empty at this point.
+                guard let response = response else {
+                    self.sendFIDO2AuthenticationError(handle: handle)
+                    return
+                }
+                self.finalizeFIDO2Authentication(handle: handle, response: response, requestId: requestId, clientDataJSON: clientDataJSON, error: nil)
+            }
+        } catch let error as NSError {
+            sendFIDO2AuthenticationError(handle: handle, errorDescription: error.localizedDescription)
+            return
```
Comment (kylehickinson) [inline Client/U2FExtensions.swift]: Redundant return

File: Client/U2FExtensions.swift [hunk 20]
```diff
@@ -0,0 +1,893 @@
...
+        let sig = response.signature.base64EncodedString()
+        
+        cleanupFIDO2Authentication(handle: handle)
+        ensureMainThread {
+            self.tab?.webView?.evaluateJavaScript("navigator.credentials.postGet('\(handle)', \(true), '\(requestId)', '\(authenticatorData)', '\(clientDataJSONString)', '\(sig)', '')", completionHandler: { _, error in
+                if error != nil {
+                    let errorDescription = error?.localizedDescription ?? U2FErrorMessages.ErrorAuthentication.rawValue
+                    log.error(errorDescription)
+                }
+            return
```
Comment (kylehickinson) [inline Client/U2FExtensions.swift]: Redundant return

File: Client/U2FExtensions.swift [hunk 21]
```diff
@@ -0,0 +1,893 @@
...
+    
+    private func sendFIDO2AuthenticationError(handle: Int, errorName: String = FIDO2ErrorMessages.NotAllowedError.rawValue, errorDescription: String = Strings.U2FAuthenticationError) {
+        cleanupFIDO2Authentication(handle: handle)
+        ensureMainThread {
+            self.tab?.webView?.evaluateJavaScript("navigator.credentials.postGet('\(handle)', \(true), '', '', '', '', '\(errorName)', '\(errorDescription)')", completionHandler: { _, error in
+                if error != nil {
+                    let errorDescription = error?.localizedDescription ?? U2FErrorMessages.ErrorAuthentication.rawValue
+                    log.error(errorDescription)
+                }
+            return
```
Comment (kylehickinson) [inline Client/U2FExtensions.swift]: Redundant return

File: Client/U2FExtensions.swift [hunk 22]
```diff
@@ -0,0 +1,893 @@
...
+            
+            self.cleanupFIDORegistration(handle: handle)
+            if requestId > -1 {
+                ensureMainThread {
+                    self.tab?.webView?.evaluateJavaScript("u2f.postLowLevelRegister(\(requestId), \(true), '\(request.version)', '\(registrationData)', '\(clientData)', \(defaultErrorCode), '')", completionHandler: { _, error in
+                        if error != nil {
+                            let errorDescription = error?.localizedDescription ?? U2FErrorMessages.ErrorRegistration.rawValue
+                            log.error(errorDescription)
+                        }
+                    return
```
Comment (kylehickinson) [inline Client/U2FExtensions.swift]: Redundant return

File: Client/U2FExtensions.swift [hunk 23]
```diff
@@ -0,0 +1,893 @@
...
+                return
+            }
+
+            ensureMainThread {
+                self.tab?.webView?.evaluateJavaScript("u2f.postRegister('\(handle)', \(true), '\(request.version)', '\(registrationData)', '\(clientData)', \(defaultErrorCode), '')", completionHandler: { _, error in
+                    if error != nil {
+                        let errorDescription = error?.localizedDescription ?? U2FErrorMessages.ErrorRegistration.rawValue
+                        log.error(errorDescription)
+                    }
+                return
```
Comment (kylehickinson) [inline Client/U2FExtensions.swift]: Im just going to stop here and say there a lots of redundant returns below that should be removed :)

File: Client/U2FExtensions.swift [hunk 24]
```diff
@@ -0,0 +1,893 @@
...
+        if message.name == "U2F", let body = message.body as? NSDictionary {
+            
+            guard let name = body["name"] as? String, let handle = body["handle"] as? Int else {
+                log.error(U2FErrorMessages.Error)
+                return
+            }
+            
+            // FIDO2 is the new webauthn API
+            // https://www.w3.org/TR/webauthn/
+            if name == "fido2-create" {
```
Comment (kylehickinson) [inline Client/U2FExtensions.swift]: Might be nice to throw these names into an enum. Either way, we should probably be using a switch statement since we're not going to get more than 1 of these flows at once

**Review by kylehickinson — CHANGES_REQUESTED** [review]

File: Client/U2FExtensions.swift [hunk 25]
```diff
@@ -0,0 +1,893 @@
...
+        fidoRegHandles.append(handle)
+        
+        guard let request = requests.first else {
+            sendFIDORegistrationError(handle: handle, requestId: id, errorCode: U2FErrorCodes.bad_request)
+            return
+        }
+    
+        fidoRegisterRequest[handle] = request
+        
+        if id > -1 {
```
Comment (iccub) [inline Client/U2FExtensions.swift]: `id >= 0` is more readable imo

File: Client/U2FExtensions.swift [hunk 26]
```diff
@@ -0,0 +1,893 @@
...
+            requestId[handle] = id
+        }
+
+        guard YubiKitManager.shared.keySession.sessionState == .open else {
```
Comment (iccub) [inline Client/U2FExtensions.swift]: In general, `guard` is not used when no value is unwrapped.
It's better to use `if YubiKitManager.shared.keySession.sessionState != .open { return }`

File: Client/U2FExtensions.swift [hunk 27]
```diff
@@ -0,0 +1,893 @@
...
+                self.sendFIDORegistrationError(handle: handle, requestId: requestId, errorCode: U2FErrorCodes.other_error)
+                return
+            }
+            guard let registrationData = response?.registrationData.websafeBase64String() else {
+                self.sendFIDORegistrationError(handle: handle, requestId: requestId, errorCode: U2FErrorCodes.other_error)
+                return
+            }
+            
+            self.cleanupFIDORegistration(handle: handle)
+            if requestId > -1 {
```
Comment (iccub) [inline Client/U2FExtensions.swift]: replace with `>= 0`

File: Client/U2FExtensions.swift [hunk 28]
```diff
@@ -0,0 +1,893 @@
...
+        requestId.removeValue(forKey: handle)
+    }
+    
+    // FIDO Authetication
+    private func requestFIDOAuthentication(handle: Int, keys: [FIDOSignRequest], id: Int) {
+        fidoSignHandles.append(handle)
+        fidoSignRequests[handle] = keys
+        fidoRequests[handle] = keys.count
+        
+        if id > -1 {
```
Comment (iccub) [inline Client/U2FExtensions.swift]: replace with `>= 0`

File: Client/U2FExtensions.swift [hunk 29]
```diff
@@ -0,0 +1,893 @@
...
+            requestId[handle] = id
+        }
+
+        guard YubiKitManager.shared.keySession.sessionState == .open else {
```
Comment (iccub) [inline Client/U2FExtensions.swift]: convert to `if != .open`

File: Client/U2FExtensions.swift [hunk 30]
```diff
@@ -0,0 +1,893 @@
...
+                }
+                guard let clientData = response?.clientData.websafeBase64String() else {
+                    self.sendFIDOAuthenticationError(handle: handle, requestId: requestId, errorCode: U2FErrorCodes.other_error)
+                    return
+                }
+                
+                self.fidoRequests[handle] = authSuccess
+                
+                self.cleanupFIDOAuthentication(handle: handle)
+                if requestId > -1 {
```
Comment (iccub) [inline Client/U2FExtensions.swift]: replace with `>= 0`

File: Client/U2FExtensions.swift [hunk 31]
```diff
@@ -0,0 +1,893 @@
...
+                    return
+                }) }
+            }
+        }
+    }
+    
+    private func sendFIDOAuthenticationError(handle: Int, requestId: Int, errorCode: U2FErrorCodes, errorMessage: String = Strings.U2FAuthenticationError) {
+        cleanupFIDOAuthentication(handle: handle)
+        
+        if requestId > -1 {
```
Comment (iccub) [inline Client/U2FExtensions.swift]: replace with `>= 0`

File: Client/U2FExtensions.swift [hunk 32]
```diff
@@ -0,0 +1,893 @@
...
+            if name == "fido2-create" {
+                guard let data = body["data"] as? String, let jsonData = data.data(using: String.Encoding.utf8) else {
+                    sendFIDO2RegistrationError(handle: handle)
+                    return
+                }
+                
+                do {
+                    let request =  try JSONDecoder().decode(WebAuthnRegisterRequest.self, from: jsonData)
+                    requestFIDO2Registration(handle: handle, request: request)
+                } catch let error as NSError {
```
Comment (iccub) [inline Client/U2FExtensions.swift]: remove `let error as NSError`

`error` variable is given for free here and no need to declare it

Do the same for other errors

**Review by iccub — CHANGES_REQUESTED** [review]
few small changes, coding style mostly

File: Client/WebAuthN/WebAuthnClientData.swift (hunk 8 above)
Comment (jumde) [reply Client/WebAuthN/WebAuthnClientData.swift:17]: I'm changing the types, `String, String, String` to `String, Data, String`. So I think we need `CodingKeys` here. Let me know if i'm missing something

File: Client/WebAuthN/WebAuthnClientData.swift (hunk 10 above)
Comment (jumde) [reply Client/WebAuthN/WebAuthnClientData.swift:36]: Used here: https://github.com/brave/brave-ios/pull/1158/files#diff-1faf1a9d0dd722c9a8d957734c58b2e6R187

File: Client/WebAuthN/WebAuthnClientData.swift (hunk 8 above)
Comment (kylehickinson) [reply Client/WebAuthN/WebAuthnClientData.swift:17]: `CodingKeys` is unrelated to types and should exist even when you don't specify it yourself (its made implicitly by the compiler) for things that conform to `Codable`.  But it seems like this struct only conforms to `Encodable`, so you will actually need this yeah

File: Client/WebAuthN/WebAuthnClientData.swift (hunk 10 above)
Comment (kylehickinson) [reply Client/WebAuthN/WebAuthnClientData.swift:36]: Ah didn't notice GitHub didn't unfold that file. Sorry!

File: Client.xcodeproj/project.pbxproj [hunk 33]
```diff
@@ -8443,6 +8492,17 @@
 					"$(BUILD_DIR)/Release$(EFFECTIVE_PLATFORM_NAME)",
 				);
 				GCC_PREPROCESSOR_DEFINITIONS = "DEBUG=1";
+				HEADER_SEARCH_PATHS = (
+					"$(SRCROOT)/ThirdParty/sqlcipher",
+					"$(inherited)",
+					/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include,
```
Comment (kylehickinson) [inline Client.xcodeproj/project.pbxproj]: This seems like a strange requirement, why do we need this?

Either way, if you really need it, it should be `$(DEVELOPER_DIR)/Toolchains/XcodeDefault.xctoolchain/usr/include`, no hard-coding paths for Xcode

File: Client.xcodeproj/project.pbxproj [hunk 34]
```diff
@@ -8895,6 +8960,17 @@
 					"$(PROJECT_DIR)/Carthage/Build/iOS",
 					"$(BUILD_DIR)/Release$(EFFECTIVE_PLATFORM_NAME)",
 				);
+				HEADER_SEARCH_PATHS = (
+					"$(SRCROOT)/ThirdParty/sqlcipher",
+					"$(inherited)",
+					/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include,
```
Comment (kylehickinson) [inline Client.xcodeproj/project.pbxproj]: Same as above, but I would guess fixing it in Xcode should fix both

File: Client.xcodeproj/project.pbxproj [hunk 35]
```diff
@@ -9225,6 +9306,17 @@
...
```
Comment (kylehickinson) [inline Client.xcodeproj/project.pbxproj]: Same as above

File: Client.xcodeproj/project.pbxproj [hunk 36]
```diff
@@ -9460,6 +9557,17 @@
...
```
Comment (kylehickinson) [inline Client.xcodeproj/project.pbxproj]: Same as above

File: Client.xcodeproj/project.pbxproj [hunk 37]
```diff
@@ -8443,6 +8492,17 @@
...
 				);
 				GCC_PREPROCESSOR_DEFINITIONS = "DEBUG=1";
+				HEADER_SEARCH_PATHS = (
+					"$(SRCROOT)/ThirdParty/sqlcipher",
+					"$(inherited)",
+					/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include,
+					"$(SRCROOT)",
+					"$(SDKROOT)/usr/include/libxml2",
+					"$(BUILD_DIR)/Release$(EFFECTIVE_PLATFORM_NAME)/include/**",
+					"$(SRCROOT)/Client/WebFilters/ShieldStats/Httpse/ThirdParty/leveldb/**",
```
Comment (kylehickinson) [inline Client.xcodeproj/project.pbxproj]: Aren't these unrelated?

**Review by kylehickinson — CHANGES_REQUESTED** [review]

**Review by kylehickinson — APPROVED** [review]
lgtm.

**Review by iccub — APPROVED** [review]
Looking good

**Comment by jhreis** [comment]
Approved from sec team via internal ticket.
